### ISSUE TITLE
Display printed output while code is still running

### ISSUE THREAD

**Issue by takluyver** [IRkernel/IRkernel#3]
The current implementation buffers printed output in a `textConnection()` until the user code has finished executing, then sends a single `stream` message with all the text. So you see no output until the cell finishes.

If we can implement an object suitable to pass to [sink](http://stat.ethz.ch/R-manual/R-patched/library/base/html/sink.html), output can be displayed as it is printed.

**Comment by flying-sheep** [comment]
[here is info about the internals](http://biostatmatt.com/R/R-conn-ints/Sinks.html)

for reference: the code for connections is in [/src/main/connections.c](https://svn.r-project.org/R/trunk/src/main/connections.c) (`do_writelines()`)

the function needs an [`Rconnection`](https://svn.r-project.org/R/trunk/src/include/R_ext/Connections.h) arg, which is a quite huge struct. do we have to instantiate one of those?

**Comment by flying-sheep** [comment]
we could maybe use [`watchout`](https://github.com/cran/evaluate/blob/master/R/watcher.r) from evaluate or similar code.

**Comment by takluyver** [comment]
See also discussion on #55.

**Comment by JanSchulz** [comment]
Not, that the python side also doesn't send any printed output during the evaluation until some `sys.stdout.flush()` is happening (which seems to be done at some place, but not always)

**Comment by takluyver** [comment]
In most cases, IPython can update output ~live. I think it flushes based on the time between successive writes to stdout.

**Comment by JanSchulz** [comment]
@takluyver I just had https://github.com/jupyter/jupyter_client/issues/7, where without the flush nothing is displayed :-)

**Comment by takluyver** [comment]
I suspect there's a corner case where there's a short burst of things being written, followed by a long gap where nothing is printed. That's harder to resolve without using threads.

**Comment by stevengj** [comment]
In Julia, we run a green thread that blocks on the file descriptor of stdout, flushes the output to a stream message when something is available to read, and then sleeps for 100ms to allow output to accumulate before blocking again.

Since this is a green thread, however, it can only wake up when something in Julia calls `yield()` (e.g. most I/O will do this).

**Comment by JanSchulz** [comment]
More examples in #230 -> it seems that this is only happening when you use a loop and the blocking process is in the loop. Also flushing the stdout doesn't matter:

```
for (i in 1:5){
    print("text")
    flush.console()
    flush(stdout())
    Sys.sleep(1)
    }
```
Could it be that evaluate caches stdout?

**Comment by JanSchulz** [comment]
Ok, looks like an evaluate issue:

```
library(evaluate)

code <- "for (i in 1:5){
print(\"text\")
message('message')
Sys.sleep(1)
}
print(\"break\")
for (i in 1:5){
  print(\"text\")
  Sys.sleep(1)
}"


l = list()
txt <- function(o, type) {
  t <- paste(o, collapse = '\n')
  t <- paste("OUT: ", t)
  l[length(l)+1] <<- paste("T: ",Sys.time()," type: ", type, "  -> ", t)
}
oh <- new_output_handler(source = identity, 
                         text = function(o) txt(o, "text"), 
                         graphics = identity,
                         message = function(o) txt(o, "message"), 
                         warning = function(o) txt(o, "warn"), 
                         error = function(o) txt(o, "error"), 
                         value = identity)



x <- evaluate(code, output_handler = oh)
l
```
[[1]]
[1] "T:  2015-12-15 17:44:36  type:  text   ->  OUT:  [1] \"text\"\n"

[[2]]
[1] "T:  2015-12-15 17:44:36  type:  message   ->  OUT:  simpleMessage in message(\"message\"): message\n\n"

[[3]]
[1] "T:  2015-12-15 17:44:37  type:  text   ->  OUT:  [1] \"text\"\n"

[[4]]
[1] "T:  2015-12-15 17:44:37  type:  message   ->  OUT:  simpleMessage in message(\"message\"): message\n\n"

[[5]]
[1] "T:  2015-12-15 17:44:38  type:  text   ->  OUT:  [1] \"text\"\n"

[[6]]
[1] "T:  2015-12-15 17:44:38  type:  message   ->  OUT:  simpleMessage in message(\"message\"): message\n\n"

[[7]]
[1] "T:  2015-12-15 17:44:39  type:  text   ->  OUT:  [1] \"text\"\n"

[[8]]
[1] "T:  2015-12-15 17:44:39  type:  message   ->  OUT:  simpleMessage in message(\"message\"): message\n\n"

[[9]]
[1] "T:  2015-12-15 17:44:40  type:  text   ->  OUT:  [1] \"text\"\n"

[[10]]
[1] "T:  2015-12-15 17:44:40  type:  message   ->  OUT:  simpleMessage in message(\"message\"): message\n\n"

[[11]]
[1] "T:  2015-12-15 17:44:41  type:  text   ->  OUT:  [1] \"break\"\n"

[[12]]
[1] "T:  2015-12-15 17:44:46  type:  text   ->  OUT:  [1] \"text\"\n[1] \"text\"\n[1] \"text\"\n[1] \"text\"\n[1] \"text\"\n"
```

**Comment by JanSchulz** [comment]
And here is the problem: https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/eval.r#L148

* each expression in a code block (=each line or in case of a loop the complete loop?) are evaluated [one after another in a loop](https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/eval.r#L148)
* messages (and errors) are handled directly when they come up thanks to `withCallingHandlers(..., message = mHandler))` ([line 149](https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/eval.r#L149)) and trigger [mhandler](https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/eval.r#L133) which calls [handle_condition](https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/eval.r#L116) which calls [handle_output](https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/eval.r#L98) which [compares the current output with the previous and sends the diff](https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/watcher.r#L28-L35) to the text handler which was defined in the `output_handler` object which we passed to evaluate.
* if no `message/...` was handled, the rest of the output will [now be handled in L152](https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/eval.r#L152)

So unless there is any way to "observe" the content of a variable (threads? something in `withCallingHandlers`?) I don't think there is any way to make this work while still using `evaluate`.

RStudio uses a `rsession.exe`, not sure how hard it is to build the kernel around that... Probably because we lose the rich display system :-/

**Comment by flying-sheep** [comment]
great detective work!

**Comment by takluyver** [comment]
We have a plan to facilitate this at some point for all kernels by running them inside a wrapper process that captures stdout/stderr at the OS level. As and when we do that, we won't need to worry about capturing printed output in R; we can just write it to the process' real stdout, and it will get captured externally.

**Comment by JanSchulz** [comment]
Just tried if there is a way to observe the inputs to a connection: http://biostatmatt.com/R/R-conn-ints.pdf So making this work in the current system would require:

* a new `callbackTextConenction` type which needs to be implemented in C?
* r must have the ability to call r code from the c code which seesm to be possible: http://stackoverflow.com/questions/7457635/calling-r-function-from-c
* Adding a patch to `evaluate` to use this...

**Comment by takluyver** [comment]
I'm not keen on the idea of maintaining anything written in C.

**Comment by JanSchulz** [comment]
Hm, wild idea: if evaluate would overwrite `flush.console()` or `flush(stdout())` wiht a version which would trigger `handle_output`, that would be a nice workaround...

**Comment by JanSchulz** [comment]
I opened https://github.com/hadley/evaluate/issues/61 for the "overwrite flush.console idea".

**Comment by JanSchulz** [comment]
https://github.com/hadley/evaluate/pull/62 has the PR which implements flushing... After (and if...) that is released, we need to add `enable_flush=TRUE` to all evaluate calls and adjust our deps...

**Comment by JanSchulz** [comment]
Unfortunately no reply yet in hadley/evaluate#62 :-( Has anybody here any connections to the devs? My last interaction with R devs left me a little scared, so I won't do a simple "ping" in the PR ... :-/

**Comment by takluyver** [comment]
Hadley Wickham is really nice, but I haven't met the person who now maintains evaluate. I've risked a ping on the PR anyway ;-)

**Comment by flying-sheep** [comment]
i know noone there :neutral_face:

**Comment by alexvpickering** [comment]
I don't know if its the same cause, but I'm getting a similar sort of bug with a loop that prints some text then draws a plot. Instead of getting the expected behaviour (text --> plot --> text --> plot --> etc.), it all gets jumbled up (text --> text --> plot --> text --> plot --> etc). Very similar effect in IPython was fixed with sys.stdout.flush()

**Comment by janschulz** [comment]
@alexvpickering Yep, this is the same cause... See also https://github.com/IRkernel/IRkernel/issues/295 for even more fun with this kind of things...

**Comment by jan-schulz-k24** [comment]
As far as I remember, this couldow be closed by adding this code before calling evaluate: https://github.com/hadley/evaluate/pull/62#issuecomment-252158714

```R
evaluate::inject_funs(
  flush.console = function() { base::flush.console(); evaluate::flush_console() },
  flush = function(con) { base::flush(con); evaluate::flush_console() }
)
```

**Comment by flying-sheep** [comment]
oh nice! anyone want to do a PR or should we do it?

**Comment by flying-sheep** [comment]
fixed in c5daa14e76920f2e6cc991eef4a0efdb82ca3e35
